src/client/network/mod.rs

```diff
@@ -394,36 +440,76 @@ enum NodeHealth {
     ///
     /// Once we've reached `healthyAt` the node is *semantically* in the ``unused`` state,
     /// other than retaining the backoff until a `healthy` request happens.
-    Unhealthy { backoff_interval: Duration, healthy_at: Instant },
+    Unhealthy { backoff: NodeBackoff, healthy_at: Instant, attempts: usize },
 
     /// When we last used the node the node acted as normal, so, we get to treat it as a healthy node for 15 minutes.
     Healthy { used_at: Instant },
 }
 
+#[derive(Copy, Clone)]
+pub(crate) struct NodeBackoff {
+    pub(crate) current_interval: Duration,
+    pub(crate) max_backoff: Duration,
+    pub(crate) min_backoff: Duration,
+    pub(crate) max_attempts: usize,
+}
+
+impl Default for NodeBackoff {
+    fn default() -> Self {
+        Self {
+            current_interval: Duration::from_millis(250),
+            max_backoff: Duration::from_secs(60 * 60),
+            min_backoff: Duration::from_millis(250),
+            max_attempts: 10,
+        }
+    }
+}
+
 impl NodeHealth {
-    fn backoff(&self) -> backoff::ExponentialBackoff {
-        const BASE_BACKOFF: Duration = Duration::from_millis(250);
-        let current_interval = match self {
-            Self::Unhealthy { backoff_interval, healthy_at: _ } => *backoff_interval,
-            _ => BASE_BACKOFF,
+    fn backoff(&self, backoff_config: NodeBackoff) -> (backoff::ExponentialBackoff, usize) {
+        // If node is already labeled Unhealthy, preserve backoff and attempt amount
+        // For new Unhealthy nodes, apply config and start attempt count at 0
+        let (node_backoff, attempts) = match self {
+            Self::Unhealthy { backoff, healthy_at: _, attempts } => (*backoff, attempts),
+            _ => (backoff_config, &0),
         };
 
-        backoff::ExponentialBackoff {
-            current_interval,
-            initial_interval: BASE_BACKOFF,
-            max_elapsed_time: None,
-            max_interval: Duration::from_secs(60 * 60),
-            ..Default::default()
+        (
+            backoff::ExponentialBackoff {
+                current_interval: node_backoff.current_interval,
+                initial_interval: node_backoff.min_backoff,
+                max_elapsed_time: None,
+                max_interval: node_backoff.max_backoff,
+                ..Default::default()
+            },
+            *attempts + 1,
+        )
+    }
+
+    pub(crate) fn mark_unhealthy(&mut self, backoff_config: NodeBackoff, now: Instant) {
+        let (mut backoff, unhealthy_node_attempts) = self.backoff(backoff_config);
+
+        // Remove node if max_attempts has been reached and max_attempts is not 0
+        if backoff_config.max_attempts > unhealthy_node_attempts && backoff_config.max_attempts != 0
+        {
+            println!("Node has reached the max amount of retries, removing from network")
```

Comment:
This shouldn't be a println, it should be a `log` message (ideally a tracing message, but we never got tracing set up), probably a `log::debug` message at that